Firestore/core/src/model/object_value.cc

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/model/object_value.h"
+
+#include <map>
+#include <set>
+
+#include "Firestore/Protos/nanopb/google/firestore/v1/document.nanopb.h"
+#include "Firestore/core/src/nanopb/message.h"
+#include "Firestore/core/src/nanopb/nanopb_util.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+MutableObjectValue::MutableObjectValue() {
+  value_->which_value_type = google_firestore_v1_Value_map_value_tag;
+  value_->map_value.fields_count = 0;
+  value_->map_value.fields =
+      nanopb::MakeArray<google_firestore_v1_MapValue_FieldsEntry>(0);
+}
+
+model::FieldMask MutableObjectValue::ToFieldMask() const {
+  return ExtractFieldMask(value_->map_value);
+}
+
+model::FieldMask MutableObjectValue::ExtractFieldMask(
+    const google_firestore_v1_MapValue& value) const {
+  std::set<FieldPath> fields;
+  for (size_t i = 0; i < value.fields_count; ++i) {
```

Comment:
The only place I see where this could make sense is after the `nested_map` declaration. I want to keep it together with the if-statement though since it belongs to the same part of the logic.